framework/0000-spa-authentication.md

```diff
@@ -0,0 +1,142 @@
+create/validate HTML forms and user credentials. Then, it relies on `Flask-Login` to create the user session and the
+cookie.
+
+Note: even if `Flask-Security` supports login via JSON request, it cannot be easily used because on login success, it
+returns as response the generated token which unfortunately cannot be stored in a secure way in a JavaScript application.
+
+### New authentication flow
+
+#### Simple login with local accounts
+
+Create a new endpoint `POST /api/login` to allow client-side application to submit an authentication request with user
+credentials. On new request, validate the payload as done for the form validation in
+[Flask-Security login()](https://github.com/mattupstate/Flask-Security/blob/develop/flask_security/views.py#L67).
+After validation, login the user in the same way via `Flask-Login` but return a different response:
+- a JSON payload with the result of the login (success or fail) and user id, to allow the client-side app to fetch the
+  user from an user endpoint
+- set a new cookie readable by JavaScript (`secure`, but not `http_only`): this new cookie must contain at least
+  a CSRF token. It can optionally contain more data useful to the client-side app, e.g. username, email and roles.
+  Both can be achieved using a JWT token, to act as CSRF token but also carry a payload.
+
+Client-side application can then access to the new cookie but cannot access the session cookie, which is protected.
+On each request with a payload, the client-side app must send the CSRF token in the request (payload or header).
+
+Any other view provided by `Flask-Security` that authenticates user and performs a render_template or redirection
+should implemented in the same way: `logout`, `register`, `psw reset`, `change psw`, etc...
+
+#### Login via OAuth
+
+Invenio implements OAuth login with the module `invenio-oauthclient`. Create a new endpoint `GET /api/oauth-providers/`
+to expose all configured OAuth providers and allow client-side apps to render a login view with all available
+authentication methods.
+
+Create a new endpoint `POST /api/login/<remote_app>` to handle the OAuth flow: on request, return a redirection to the
+OAuth provider to perform the first part of the OAuth flow setting as callback parameter the backend authorized
+url `/api/authorized/`.
+On callback called, `invenio-oauthclient` should complete the OAuth flow as already implemented, but perform the same
+operations of double cookie creation as described before. It should then redirect back to the client-side app with a
+success/failure status.
+
+### CSRF
+
+Any HTTP request that sends a payload and relies on cookies authentication should be protected from CSRF attacks.
+The backend, for cookie authenticated requests only (not token based ones), should require and validate the CSRF token
+in the payload or as HTTP header.
+
+All Invenio REST endpoints should be reviewed to require CSRF token.
+Note: this is a **backward incompatible** change!
```

Comment:
Not 100% sure about that, but I think we could keep backward compatibility since we're talking about a completely new method for auth where a JWT is sent via cookies along with the random CSRF number (e.g. this can bee detected at the beginning of the request by some `app.before_request`). In that case, the CSRF validation doesn't have to happen on each form, but on a pre-processor, meaning we don't have to go through all of the views/forms and mark them.

Reply:
Indeed, it was implemented like this, changing the RFC.